### PR TITLE
General fixes for Base

### PR DESCRIPTION
**Pull request by theiostream**
This pull request provides a series of general fixes to Base for increased compatibility with newer OSX APIs used by WebKit.

The commit logs add more insight into the individual changes done to the project.

### CODE REVIEW

**Comment by ivucica** [comment]
It might be interesting to have smaller pull requests in the future, consisting of just a few, very related commits?

I'm sorry for not noticing this PR earlier. I think @rfm would also be a better person to do the review.

However, for starters, I guess that since we /do/ have Travis set up, maybe the review should wait until you make the build go green?

File: Headers/Foundation/NSCoder.h [hunk 1]
```diff
@@ -363,6 +363,9 @@ extern "C" {
  * Decodes an NSInteger associated with the key.
  */
 - (NSInteger) decodeIntegerForKey: (NSString *)key;
+
+
+@property (nonatomic, assign) BOOL requiresSecureCoding;
```
Comment (fredkiefer) [inline Headers/Foundation/NSCoder.h]: Please replace this with a getter and a setter.

File: Headers/Foundation/NSDictionary.h [hunk 2]
```diff
@@ -67,8 +67,9 @@ extern "C" {
 			     indent: (NSUInteger)level;
 
 #if OS_API_VERSION(MAC_OS_X_VERSION_10_6, GS_API_LATEST)
-DEFINE_BLOCK_TYPE(GSKeysAndObjectsEnumeratorBlock, void,
-  GS_GENERIC_TYPE_F(KeyT,id<NSCopying>), GS_GENERIC_TYPE(ValT), BOOL*);
+/*DEFINE_BLOCK_TYPE(GSKeysAndObjectsEnumeratorBlock, void,
```
Comment (fredkiefer) [inline Headers/Foundation/NSDictionary.h]: This is wrong and you already know it. But why did you change it anyway? This breaks your pull request completely.

File: Headers/Foundation/NSFileManager.h [hunk 3]
```diff
@@ -230,6 +233,10 @@ typedef NSUInteger NSDirectoryEnumerationOptions;
 + (NSFileManager*) defaultManager;
 
 #if OS_API_VERSION(MAC_OS_X_VERSION_10_5, GS_API_LATEST)
+@property (assign) id<NSFileManagerDelegate> delegate;
```
Comment (fredkiefer) [inline Headers/Foundation/NSFileManager.h]: Again, please provide a getter and a setter.

File: Headers/Foundation/NSFileManager.h [hunk 4]
```diff
@@ -659,6 +669,68 @@ GS_EXPORT NSString* const NSFileSystemFreeNodes;
 - (NSUInteger) fileSystemFileNumber;
 @end
 
+@protocol NSFileManagerDelegate <NSObject>
```
Comment (fredkiefer) [inline Headers/Foundation/NSFileManager.h]: Please have a look at NSMetaData.h on how to do this in a way old gcc can cope with.

@protocol NSMetadataQueryDelegate
#if OS_API_VERSION(MAC_OS_X_VERSION_10_6,GS_API_LATEST) && GS_PROTOCOLS_HAVE_OPTIONAL
@optional
#else
@end
@interface NSObject (NSMetadataQueryDelegate)
#endif // GS_PROTOCOLS_HAVE_OPTIONAL

File: Headers/Foundation/NSHTTPCookie.h [hunk 5]
```diff
@@ -204,13 +204,13 @@ extern NSString * const NSHTTPCookieVersion; /** Obtain cookie version */
  * Returns whether the receiver should only be sent over
  * secure connections.
  */
-- (BOOL) isSecure;
+@property (readonly, getter=isSecure) BOOL secure;
```
Comment (fredkiefer) [inline Headers/Foundation/NSHTTPCookie.h]: Please restore the old state.

File: Headers/Foundation/NSHTTPCookie.h [hunk 6]
```diff
@@ -204,13 +204,13 @@ extern NSString * const NSHTTPCookieVersion; /** Obtain cookie version */
...
 
 /**
  * Returns whether the receiver should be destroyed at the end of the
  * session.
  */
-- (BOOL) isSessionOnly;
+@property (readonly, getter=isSessionOnly) BOOL sessionOnly;
```
Comment (fredkiefer) [inline Headers/Foundation/NSHTTPCookie.h]: Please restore the old state.

File: Headers/Foundation/NSHTTPCookie.h [hunk 7]
```diff
@@ -247,6 +247,10 @@ extern NSString * const NSHTTPCookieVersion; /** Obtain cookie version */
  */
 - (NSUInteger) version;
 
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_6,GS_API_LATEST)
+@property (readonly, getter=isHTTPOnly) BOOL HTTPOnly;
```
Comment (fredkiefer) [inline Headers/Foundation/NSHTTPCookie.h]: Again a getter will be needed.

File: Headers/Foundation/NSProcessInfo.h [hunk 8]
```diff
@@ -83,6 +101,10 @@ enum {
  */
 + (NSProcessInfo*) processInfo;
 
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_6,GS_API_LATEST) 
+@property (readonly) NSUInteger systemUptime;
```
Comment (fredkiefer) [inline Headers/Foundation/NSProcessInfo.h]: Please add a getter method instead.

File: Headers/Foundation/NSURL.h [hunk 9]
```diff
@@ -149,6 +171,8 @@ GS_EXPORT NSString* const NSURLFileScheme;
 - (id) initWithString: (NSString*)aUrlString
 	relativeToURL: (NSURL*)aBaseUrl;
 
+@property (readonly, getter=isFileURL) BOOL fileURL;
```
Comment (fredkiefer) [inline Headers/Foundation/NSURL.h]: Please add a getter method.

File: Headers/Foundation/NSURLConnection.h [hunk 10]
```diff
@@ -86,6 +86,24 @@ extern "C" {
  */
 - (id) initWithRequest: (NSURLRequest *)request delegate: (id)delegate;
 
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_8,GS_API_LATEST)
+@property (readonly, copy) NSURLRequest *currentRequest;
```
Comment (fredkiefer) [inline Headers/Foundation/NSURLConnection.h]: getter and setter...

File: Headers/Foundation/NSURLProtectionSpace.h [hunk 11]
```diff
@@ -129,6 +138,10 @@ authenticationMethod: (NSString *)authenticationMethod;
  */
 - (BOOL) receivesCredentialSecurely;
 
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_6,GS_API_LATEST) && GS_API_VERSION( 11300,GS_API_LATEST)
+@property (readonly, copy) NSArray *distinguishedNames;
```
Comment (fredkiefer) [inline Headers/Foundation/NSURLProtectionSpace.h]: Getter...

File: Source/NSCoder.m [hunk 12]
```diff
@@ -397,6 +408,12 @@ - (id) decodeObjectForKey: (NSString*)aKey
   return nil;
 }
 
+- (id) decodeObjectOfClass: (Class)cls forKey: (NSString *)aKey
```
Comment (fredkiefer) [inline Source/NSCoder.m]: Please also add the method decodeObjectOfClasses:forKey: as that is the more basic one, which should actually be implemented.

File: Source/NSFileManager.m [hunk 13]
```diff
@@ -2032,6 +2026,7 @@ - (NSDictionary*) fileSystemAttributesAtPath: (NSString*)path
 
   if (!GetVolumePathNameW(lpath, volumePathName, 128))
     {
+      *error = [NSError _last];
```
Comment (fredkiefer) [inline Source/NSFileManager.m]: Shouldn't you check whether error is not NULL?

File: Source/NSHTTPCookie.m [hunk 14]
```diff
@@ -59,6 +59,7 @@ after this (See GSRangeOfCookie).
 NSString * const NSHTTPCookieSecure = @"Secure";
 NSString * const NSHTTPCookieValue = @"Value";
 NSString * const NSHTTPCookieVersion = @"Version";
+static NSString * const NSHTTPCookieHTTPOnly = @"HTTPOnly";
```
Comment (fredkiefer) [inline Source/NSHTTPCookie.m]: I would not use a variable name starting with NS for a static variable. Apple uses this name space and we might get a clash later on.

File: Source/NSHTTPCookie.m [hunk 15]
```diff
@@ -736,7 +745,18 @@ static inline id parseUnquotedString(pldata *pld, char endChar)
 	}
       else
 	{
+	  unsigned int oldpos = pld->pos, keyvalpos;
```
Comment (fredkiefer) [inline Source/NSHTTPCookie.m]: Please use a separate line for each variable.

File: Source/NSURLConnection.m [hunk 16]
```diff
@@ -226,12 +257,36 @@ - (id) initWithRequest: (NSURLRequest *)request delegate: (id)delegate
 	initWithRequest: this->_request
 	cachedResponse: nil
 	client: (id<NSURLProtocolClient>)self];
-      [this->_protocol startLoading];
+      if (immediately)
+        {
+          this->_started = YES;
```
Comment (fredkiefer) [inline Source/NSURLConnection.m]: Why don't you call the method start here?

File: Source/NSURLConnection.m [hunk 17]
```diff
@@ -452,6 +507,11 @@ - (void) URLProtocol: (NSURLProtocol *)protocol
 	initWithRequest: this->_request
 	cachedResponse: nil
 	client: (id<NSURLProtocolClient>)self];
+
```
Comment (fredkiefer) [inline Source/NSURLConnection.m]: Again, wouldn't calling the method start be more consistent here? E.g. that would set the new _started instance variable.

File: Source/NSURLConnection.m [hunk 18]
```diff
@@ -476,5 +536,80 @@ - (void) URLProtocol: (NSURLProtocol *)protocol
...
+    }
+  
+  if (this->_customOperationQueue)
+    {
+      NSLog(@"you cannot use both run loops and an operation queue");
+      return;
+    }
+  
+  NSMutableArray *modes = [this->_customRunLoops objectForKey: loop];
+  if ([this->_customRunLoops objectForKey: loop])
```
Comment (fredkiefer) [inline Source/NSURLConnection.m]: Why don't you use your local variable modes here?

File: Source/NSURLCredential.m [hunk 19]
```diff
@@ -110,6 +110,9 @@ - (id) initWithUser: (NSString *)user
     }
   if ((self = [super init]) != nil)
     {
+      if (persistence == NSURLCredentialPersistenceSynchronizable)
+	persistence = NSURLCredentialPersistencePermanent;
```
Comment (fredkiefer) [inline Source/NSURLCredential.m]: Please use {} even for single line if statements.

File: Source/NSURLProtocol.m [hunk 20]
```diff
@@ -228,10 +228,14 @@ - (void) close
 {
   [ip setDelegate: nil];
   [op setDelegate: nil];
+
+  // FIXME: should we add support for other run loops
+  // or to the operation queue here?
```
Comment (fredkiefer) [inline Source/NSURLProtocol.m]: Yes! And maybe call this close method from dealloc?

File: Source/NSURLProtocol.m [hunk 21]
```diff
@@ -1802,10 +1979,32 @@ - (void) stopLoading
     {
       [this->input setDelegate: nil];
       [this->output setDelegate: nil];
-      [this->input removeFromRunLoop: [NSRunLoop currentRunLoop]
-			     forMode: NSDefaultRunLoopMode];
-      [this->output removeFromRunLoop: [NSRunLoop currentRunLoop]
-			      forMode: NSDefaultRunLoopMode];
+      if (this->customOperationQueue)
```
Comment (fredkiefer) [inline Source/NSURLProtocol.m]: Could you please try to reduce the duplication of code in this class? To me it seems like you had to copy similar bits over and over again.

File: Headers/Foundation/NSDate.h [hunk 22]
```diff
@@ -29,6 +29,8 @@
 extern "C" {
 #endif
 
+GS_EXPORT NSString * const NSSystemClockDidChangeNotification;
```
Comment (fredkiefer) [inline Headers/Foundation/NSDate.h]: You need to add a definition of this value somewhere.

**Comment by fredkiefer** [comment]
I added a few commets inline.

File: Headers/Foundation/NSHTTPCookie.h (hunk 6 above)
Comment (ivucica) [inline Headers/Foundation/NSHTTPCookie.h]: To clarify: if the `@property` syntax is required for compatibility, please provide it as an alternative and guarded by appropriate feature guards.

Alternatively, provide `- (BOOL) isSessionOnly` plus `- (BOOL) sessionOnly` as a wrapper around it.

`@property` is really breaking the 3.4.5 requirement which we outlined in the mailing list.

File: Headers/Foundation/NSDictionary.h (hunk 2 above)
Comment (theiostream) [inline Headers/Foundation/NSDictionary.h]: Was a temporary (and very quick) fix to avoid a casting hell error when building WebKit, and I kept it as a commit as a reminder that I'd have to look at it later. Should not have included it in this pull request.

File: Source/NSFileManager.m (hunk 13 above)
Comment (theiostream) [inline Source/NSFileManager.m]: I think there are other places in the code that do not check for this, but I'll investigate further.

In any case, having `*error` set to `NULL` is not wrong (although it would be quite misleading). In this case, what should we do? Generate a new `NSError` right here? What should it contain?

**Comment by rfm** [comment]
> On 28 Jun 2017, at 13:09, Daniel Ferreira <notifications@github.com> wrote:
> 
> @theiostream commented on this pull request.
> 
> In Source/NSFileManager.m:
> 
> > @@ -2032,6 +2026,7 @@ - (NSDictionary*) fileSystemAttributesAtPath: (NSString*)path
>  
>    if (!GetVolumePathNameW(lpath, volumePathName, 128))
>      {
> +      *error = [NSError _last];
> 
> I think there are other places in the code that do not check for this, but I'll investigate further.
> 
> In any case, having *error set to NULL is not wrong (although it would be quite misleading). In this case, what should we do? Generate a new NSError right here? What should it contain?

I think Fred was saying that you have a potential null pointer dereference.
So the code ought to be:

 if (!GetVolumePathNameW(lpath, volumePathName, 128))
     {
       if (NULL != error)
	{
	  *error = [NSError _last];
        }

Generally if a method has pointer arguments used to return valuers, we ought to check to see if those pointers are null and do somethjing sensible (either quietly refrain from assigning to the address or raise an exception ... whatever OSX does).

**Comment by theiostream** [comment]
Hi @fredkiefer, thanks for the comments. I replied in those I had something to say about; otherwise, I agree completely.

I think I'd mentioned in a prior e-mail that I knew `@property`s were still an unsolved problem (I hope I'd spare you the effort of pointing out every misuse of them). @rfm's comment did not show up in the issue tracker, so I'll copy it for visibility:

```
While we definitely want to provide explicit getter/setter methods for any properties declared in new Apple headers, that's not really sufficient.  We also need to conditionally use the @property declaration.

[...]

I'm wondering what the best way to proceed is:
1. change the configure script for the base library to check to see if the current compiler supports @property and define GS_HAS_DECLARED_PROPERTIES in GSConfig.h so all our headers can use it.
2. change library-combo.make to define GS_HAS_DECLARED_PROPERTIES if the next generation (ng) librarary is selected (for new ObjC features with clang) and don't use properties otherwise.
```

How do gccs with new Objective-C 2.0 runtimes enter this picture?

File: Source/NSFileManager.m (hunk 13 above)
Comment (theiostream) [inline Source/NSFileManager.m]: Oh, nevermind. I just got where the problem is (thanks to @rfm's insight). Thanks for the heads-up! :)

**Comment by theiostream** [comment]
A separate comment about `NSURLConnection`/`NSURLProtocol` stuff: I agree with all of the issues you mentioned (in fact, I realized them myself as I went on coding it). I'm determined to fix them, but before I continued any work on that I wanted your two cents on that compatibility matter with `NSURLQueue` I mentioned in some prior emails (copying here for visibility):

```
So it seems like -setDelegateQueue: not only gets delegate calls
redirected to that queue. It *also* guarantees that even if the thread
where -[NSURLConnection start] was called on dies, the delegate calls
will still be placed on the queue. Unfortunately, this seems hard to
conciliate with our reliance on NSStream's -handleEvent: model, which
requires -[NSURLConnection start] to be run on a thread with a run
loop running.

While this does seem complicated, the only workaround I can think of
not to break compatibility with this Apple guarantee is do everything
-[NSURLProtocol startLoading] does in a new thread managed by us that
will always have the run loop running. In case we're using
NSOperationQueue, we could easily add operations to it from there; in
case we're not, we could keep a reference to the thread that called
-[NSURLConnection init] or -[NSURLConnection start] originally and
dispatch the delegate selectors to them. This way we guarantee that if
the thread that called -[NSURLConnection start] dies, we'll still get
the NSStream events. Of course, I may be delirious here, so please
correct me if that's a stupid way to go about this.
```

Again, thanks for the thorough review.

**Comment by theiostream** [comment]
Hm, @rfm, when Fred pointed me to the `GS_PROTOCOLS_HAVE_OPTIONAL` I couldn't help but think our definition of `GS_HAS_DECLARED_PROPERTIES` could have the same style: https://github.com/gnustep/libs-base/pull/8/commits/8ffc352abed8ea156bff7fcf9f7e253f2dad6656

What do you think?

@fredkiefer the branch has been updated fixing everything you pointed out (hopefully). I removed the two WIP commits.

**Comment by rfm** [comment]
> On 28 Jun 2017, at 15:31, Daniel Ferreira <notifications@github.com> wrote:
> 
> Hm, @rfm, when Fred pointed me to the GS_PROTOCOLS_HAVE_OPTIONAL I couldn't help but think our definition of GS_HAS_DECLARED_PROPERTIES could have the same style: 8ffc352
> 
> What do you think?

Yes, I actually think that looks better than either of my suggestions.

File: Headers/Foundation/NSCoder.h [hunk 23]
```diff
@@ -364,6 +367,19 @@ extern "C" {
  */
 - (NSInteger) decodeIntegerForKey: (NSString *)key;
 #endif
+
+#if OS_API_VERSION(MAC_OS_X_VERSION_10_8, GS_API_LATEST)
+
+#if GS_HAS_DECLARED_PROPERTIES
+@property (nonatomic, assign) BOOL requiresSecureCoding;
+#else
+- (BOOL) requiresSecureCoding;
```
Comment (fredkiefer) [inline Headers/Foundation/NSCoder.h]: I think you will also need a setter, not just a getter method.

File: Headers/Foundation/NSFileManager.h [hunk 24]
```diff
@@ -230,6 +233,14 @@ typedef NSUInteger NSDirectoryEnumerationOptions;
 + (NSFileManager*) defaultManager;
 
 #if OS_API_VERSION(MAC_OS_X_VERSION_10_5, GS_API_LATEST)
+#if GS_HAS_DECLARED_PROPERTIES
+@property (assign) id<NSFileManagerDelegate> delegate;
+#else
+- (id<NSFileManagerDelegate>) delegate;
```
Comment (fredkiefer) [inline Headers/Foundation/NSFileManager.h]: Again the setter seems to be missing.

**Comment by theiostream** [comment]
Gah, I've just noticed that (apparently starting some days ago) GitHub decided to sort commits not by their `git log` order but by their date, which is totally braindead. It makes it seem that the `GS_HAS_DECLARED_PROPERTIES` definition, on top of which the other commits are built, comes after because I created it today, even though it was created through a rebase.

This probably makes reviewing individual commits (as usually makes sense -- see the git project itself, for instance) hell through this web UI. Knowing that I will continue to rebase my changes, let me know if there is any better way for you to read my commits.

**Comment by ivucica** [comment]
Looking at Github review UI for anything but the smallest PRs, I wish I could justify setting up Gerrit and getting people to use its approach to submitting changes. This in particular is a disaster.

**Comment by ivucica** [comment]
Or if Github had a more decent diff between pushes, which breaks especially if you have rebases.

I think you could have merged this as well, but I'll do it now.

**Comment by theiostream** [comment]
@fredkiefer will make sure to do that next time. Thanks for the heads-up.
